pubsub/azure/servicebus/servicebus.go

```diff
@@ -84,6 +86,15 @@ func parseAzureServiceBusMetadata(meta pubsub.Metadata) (metadata, error) {
 		}
 	}
 
+	m.NumConcurrentConsumers = runtime.NumCPU()
```

Comment:
As mentioned in the issue, I think we should by default dispatch on a new goroutine for every message (instead of limiting to number of CPUs, as Go can schedule hundreds of thousands of Goroutines per core), and use `numConcurrentConsumers` to limit.

Reply:
Ok makes sense. The way the sdk is designed means a call to `Receive()` is blocking with an internal message pump which pushes the messages onto a channel which are then pulled and handled by a function you provide on a separate go routine. [Their example for handling concurrency](https://github.com/Azure/azure-service-bus-go/blob/1839bfc2a41a03ac130d55a1715188d3c0552f07/queue_examples_test.go#L113) works like this, where you create `n` listeners. I'll have a look at this tomorrow and see if we can handle this slightly differently to create a go routine per message. Welcome thoughts.

Reply:
Looking at that example, we might not need lines 125 - 136.

We can create a new goroutine from the handler, something like this: (very much pseudo code)

```go
var concurrentHandler servicebus.HandlerFunc = func(ctx context.Context, msg *servicebus.Message) error {
		go func() {
                      ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		      defer cancel()

                         // send message to the user app
                       err := handler(msg)
                       if err == nil {
                          msg.Complete(ctx)
                      }
                 }()
		return nil
}
```

Reply:
Also, concurrency limit can be a global setting for Pub/Sub in Dapr instead of implemented individually.. see here:

https://github.com/dapr/dapr/issues/940#issuecomment-571789035

Reply:
Thanks for the feedback @yaron2 - interested to understand the "global setting" approach a bit more. Are you referring to adding a property to the `Config` struct in `dapr/runtime/config.go` that differs to `MaxConcurrency` such as `PubSubMaxConsumers` and then passing the runtime config into the pubsub in `initPubSub` , or did you have something else in mind?

Reply:
It'll be another `spec.metadata` property on Pub Sub components, named `concurrencyLimit`.
The runtime will then enforce it in `daprd` for every pub/sub implementation.

Reply:
So my suggestion is to remove concurrency limits completely from the Azure Service bus code.

Reply:
Is it best to close this PR then?

Reply:
Well, looking into the ASB implementation now, unfortunately we should keep this PR, as ASB waits for the handler function to finish before executing a new handler, so it itself enforces concurrency, meaning the only way to allow for more than one execution is to create N goroutines.

Concurrency limit will still be handled in the Dapr runtime as I outlined above, but that's under the expectation that individual implementations by default do not enforce concurrency.

The question now is how to calculate the default concurrency limit.
Since in Go gourtines != threads and the Go scheduler is smart enough to grow/shrink the stack size and reallocate routines on threads, I think we can safely start with sane defaults of 1000 goroutines per CPU core.

Reply:
I would still leave the metadata property for an option for power users.

Reply:
Just to clarify - for the ASB impl we keep an additional metadata property (numConcurrentConsumers or equivalent) and use a default of 1000 per core. There will be an additional mechanism in the runtime that will use global property limitConcurrency to limit other pubsubs but will be ignored by this impl. Feels like this will need to be very clear in the docs to stop people tripping up.

Reply:
The Dapr runtime expects implementations to fire as many events in parallel as they can, with respects to individual configurations.

The Dapr runtime setting to limit will always take the rate of incoming events from a pub/sub components and limit it, regardless of what happens in the implementation itself.

The components are unaware of what the Dapr runtimes does or not. the `numConcurrentConsumers` in ASB is not to limit concurrency, its essentially there to work around the fact that the ASB itself is serially executing a handler message per worker.

Reply:
Ok, so in the current implementation - the handler is effectively async (which bypasses error handling etc. in the sdk) but the message iterator is serial. Is it best we revert to this PRs original implementation which ran 'n' (in this case we think 1000 per core) workers each which processing messages serially? Maybe rename `numConcurrentConsumers` to `numListeners` and then leave the throttling to the Dapr runtime.

Reply:
Ok so; `limitConcurrency` in runtime will effectively throttle the number of concurrent requests dapr makes to your app. `numConcurrentConsumers` would effectively be a limit on the number of concurrent handlers (and thus goroutines) running in the pubsub component. Is that about right? Maybe renamed `numConcurrentConsumers` to `numConcurrentHandlers`. We leave that without a default (leaving it to the go runtime) and allow power users to set it to effectively control the number of goroutines the pubsub component itself can spin up.

Reply:
That is correct. +1 on suggested property names.

Reply:
ok great I'll resolve this thread - any additional comments can be made on a new review